contracts/Tendermint-ORU.sol

```diff
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.6.0 <8.0.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+
+/// @notice Submission of remote chain block header.
+struct HeaderSubmission {
+    // Block header
+    Header header;
+    // Submitter
+    address payable submitter;
+    // Ethereum block number submission was made
+    uint256 blockNumber;
+    // Simple hash of previous block's commit in ABI encoded format
+    bytes32 lastCommitHash;
+}
+
+/// @notice Remote chain block header.
+/// @dev https://github.com/lazyledger/lazyledger-specs/blob/master/specs/data_structures.md#header
+struct Header {
+    uint64 height;
+    uint64 timestamp;
+    bytes32 lastBlockID;
+    bytes32 lastCommitRoot;
+    bytes32 consensusRoot;
+    bytes32 stateCommitment;
+    bytes32 availableDataRoot;
+    bytes32 proposerAddress; // Note: this needs to be converted to a 20-byte address when used
+}
+
+/// @notice Compact ECDSA signature.
+/// @dev https://github.com/lazyledger/lazyledger-specs/blob/master/specs/data_structures.md#signature
+struct Signature {
+    bytes32 r;
+    bytes32 vs;
+}
+
+/// @notice Tendermint signature.
+/// @dev https://github.com/lazyledger/lazyledger-specs/blob/master/specs/data_structures.md#commitsig
+struct CommitSig {
+    uint8 blockIDFlag;
+    bytes32 validatorAddress; // Note: this needs to be converted to a 20-byte address when used
```

Comment:
I don't see any places where this is used/consumed yet. I assume that it's fairly easy to truncate a bytes32 in solidity?

Reply:
It's not used yet because it only needs to be checked when doing a fraud proof (which would involve actually verifying all the `CommitSig`s), and that's a TODO.